src/Lean/Compiler/LCNF/JoinPoints.lean

```diff
@@ -236,6 +286,235 @@ where
 
 end JoinPointFinder
 
+namespace JoinPointContextExtender
+
+open ScopeM
+
+/--
+The context managed by `ExtendM`.
+-/
+structure ExtendContext where
+  /--
+  The `FVarId` of the current join point if we are currently inside one.
+  -/
+  currentJp? : Option FVarId := none
+  /--
+  The list of valid candidates for extending the context. This will be
+  all `let` and `fun` declarations as well as all `jp` parameters up
+  until the last `fun` declaration in the tree.
+  -/
+  candidates : FVarIdSet := {}
+
+/--
+The state managed by `ExtendM`.
+-/
+structure ExtendState where
+  /--
+  A map from join point `FVarId`s to a respective map from free variables
+  to `Param`s. The free variables in this map are the once that the context
+  of said join point will be extended by by passing in the respective parameter.
+  -/
+  fvarMap : HashMap FVarId (HashMap FVarId Param) := {}
```

Comment:
It is usually hard do use this kind of `map` linearly and ensure we are getting destructive updates. For example:
```
    let mut translator := (←get).fvarMap.find! currentJp
    ...
         -- The following update is not destructive since `(<- get).fvarMap` still contains a reference to the map
          translator := translator.insert fvar newParam
```
This is probably not a big deal here since the nested map should be small, and copy is not a big deal.